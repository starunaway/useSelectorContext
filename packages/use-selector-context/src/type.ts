import { Context } from 'react';

const Z = Symbol.for('ZContext');

export interface ZContext<Value> {
  [Z]: {};
  Provider: Context<Value>['Provider'];
  Consumer: null;
}

export type ZContextValue<Value = unknown> = {
  getSnapshot: () => Value;
  subscribe: (listener: Listener) => () => void;
};

export type Listener = () => void;

export type EqualityFn<T> = (a: T, b: T) => boolean;
