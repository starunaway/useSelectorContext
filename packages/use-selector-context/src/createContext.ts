import React, { ReactNode, Provider, useRef, useLayoutEffect } from 'react';
import { Listener, ZContext, ZContextValue } from './type';

function createSubscribe(listeners: Set<Listener>) {
  return function subscribe(listener: Listener) {
    listeners.add(listener);
    return function unsubscribe() {
      listeners.delete(listener);
    };
  };
}

function createProvider<Value>(originProvider: Provider<ZContextValue<Value>>) {
  const ContextProvider = ({ value, children }: { value: Value; children: ReactNode }) => {
    const contextValue = useRef<ZContextValue<Value>>();

    const realValue = useRef<Value>(value);
    const listeners = useRef(new Set<Listener>());

    if (!contextValue.current) {
      contextValue.current = {
        getSnapshot: () => realValue.current,
        subscribe: createSubscribe(listeners.current),
      };
    }

    useLayoutEffect(() => {
      realValue.current = value;
      listeners.current.forEach((listener) => listener());
    }, [value]);

    return React.createElement(originProvider, { value: contextValue.current }, children);
  };

  return ContextProvider as ZContext<Value>['Provider'];
}

export default function createContext<Value>(defaultValue: Value) {
  const contextValue: ZContextValue<Value> = {
    getSnapshot: () => defaultValue,
    subscribe: {} as any,
  };

  const context = React.createContext<ZContextValue<Value>>(contextValue);

  (context as any).Provider = createProvider(context.Provider);

  return context as unknown as ZContext<Value>;
}
