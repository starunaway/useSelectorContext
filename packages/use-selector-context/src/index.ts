export { default as useContext } from './hooks/useContext';
export { default as useGetSnapshot } from './hooks/useGetSnapshot';
export { default as createContext } from './createContext';
export * from './type';
export { default as shallowEqual } from './utils/shallowEqual';
