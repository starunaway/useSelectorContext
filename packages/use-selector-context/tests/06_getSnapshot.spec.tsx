import React, { useState, StrictMode, useRef, memo } from 'react';
import { afterEach, describe, expect, it } from 'vitest';
import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import { createContext, useContext, useGetSnapshot } from '../src';

describe('04_withContext', () => {
  afterEach(cleanup);

  it('counter', () => {
    const initialState = {
      count1: 0,
      setCount1: (() => {}) as React.Dispatch<React.SetStateAction<number>>,
      count2: 0,
      setCount2: (() => {}) as React.Dispatch<React.SetStateAction<number>>,
    };
    const context = createContext(initialState);

    const Counter1 = () => {
      const count = useContext(context, (v) => v.count1);
      const setState = useContext(context, (v) => v.setCount1);
      const increment = () => setState((s) => s + 1);

      return (
        <div>
          <span data-testid="counter1">{count}</span>
          <button data-testid="button" onClick={increment}>
            +1
          </button>
        </div>
      );
    };

    const useGetAppSnapshot = useGetSnapshot.withContext(context);
    const MemoCounterA = memo(() => {
      const getSnapshot = useGetAppSnapshot();
      const [snapShot, setSnapShot] = useState<ReturnType<typeof getSnapshot> | null>(null);

      const renderCount = useRef(0);
      renderCount.current += 1;
      return (
        <div>
          <span data-testid="snapShotA">{snapShot?.count1}</span>
          <button data-testid="btnA" onClick={() => setSnapShot(getSnapshot())}>
            getSnapshot
          </button>
          <span data-testid="memocounterA">{renderCount.current}</span>
        </div>
      );
    });

    const MemoCounterB = memo(() => {
      const getSnapshot = useGetSnapshot(context);
      const [snapShot, setSnapShot] = useState<ReturnType<typeof getSnapshot> | null>(null);

      const renderCount = useRef(0);
      renderCount.current += 1;
      return (
        <div>
          <span data-testid="snapShotB">{snapShot?.count1}</span>
          <button data-testid="btnB" onClick={() => setSnapShot(getSnapshot())}>
            getSnapshot
          </button>
          <span data-testid="memocounterB">{renderCount.current}</span>
        </div>
      );
    });

    const App = () => {
      const [count1, setCount1] = useState(0);
      const [count2, setCount2] = useState(0);
      return (
        <StrictMode>
          <context.Provider
            value={{
              count1,
              setCount1,
              count2,
              setCount2,
            }}
          >
            <Counter1 />
            <MemoCounterA />
            <MemoCounterB />
          </context.Provider>
        </StrictMode>
      );
    };

    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId('button'));
    expect(screen.getByTestId('counter1').textContent).toEqual('1');
    expect(screen.getByTestId('memocounterA').textContent).toEqual('1');
    expect(screen.getByTestId('memocounterB').textContent).toEqual('1');
    expect(screen.getByTestId('snapShotA').textContent).toEqual('');
    expect(screen.getByTestId('snapShotB').textContent).toEqual('');

    fireEvent.click(screen.getByTestId('btnA'));
    expect(screen.getByTestId('memocounterA').textContent).toEqual('3');
    expect(screen.getByTestId('memocounterB').textContent).toEqual('1');
    expect(screen.getByTestId('snapShotA').textContent).toEqual('1');
    expect(screen.getByTestId('snapShotB').textContent).toEqual('');

    fireEvent.click(screen.getByTestId('btnB'));
    expect(screen.getByTestId('memocounterA').textContent).toEqual('3');
    expect(screen.getByTestId('memocounterB').textContent).toEqual('3');
    expect(screen.getByTestId('snapShotA').textContent).toEqual('1');
    expect(screen.getByTestId('snapShotB').textContent).toEqual('1');
    expect(container).toMatchSnapshot();
  });
});
