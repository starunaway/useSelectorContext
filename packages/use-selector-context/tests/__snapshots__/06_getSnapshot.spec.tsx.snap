// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`04_withContext > counter 1`] = `
<div>
  <div>
    <span
      data-testid="counter1"
    >
      0
    </span>
    <button
      data-testid="button"
    >
      +1
    </button>
  </div>
  <div>
    <span
      data-testid="snapShotA"
    />
    <button
      data-testid="btnA"
    >
      getSnapshot
    </button>
    <span
      data-testid="memocounterA"
    >
      1
    </span>
  </div>
  <div>
    <span
      data-testid="snapShotB"
    />
    <button
      data-testid="btnB"
    >
      getSnapshot
    </button>
    <span
      data-testid="memocounterB"
    >
      1
    </span>
  </div>
</div>
`;

exports[`04_withContext > counter 2`] = `
<div>
  <div>
    <span
      data-testid="counter1"
    >
      1
    </span>
    <button
      data-testid="button"
    >
      +1
    </button>
  </div>
  <div>
    <span
      data-testid="snapShotA"
    >
      1
    </span>
    <button
      data-testid="btnA"
    >
      getSnapshot
    </button>
    <span
      data-testid="memocounterA"
    >
      3
    </span>
  </div>
  <div>
    <span
      data-testid="snapShotB"
    >
      1
    </span>
    <button
      data-testid="btnB"
    >
      getSnapshot
    </button>
    <span
      data-testid="memocounterB"
    >
      3
    </span>
  </div>
</div>
`;
